Add support for Scoring functionality

Cheng20010201 opened the pull request:
`Score` data should be stored in user's local data file, not the database file. The database file (`src/main/resources/database/JapansesDatabase.json`) is meant to be effectively read-only to ensure data integrity.

1. Create a `Score.java` under package `model` to represent an entry of quiz attempt result in user data storage.
2. Relevant testsdocumentation.
3. Actual storing mechanism
4. Primitive GUI rendering
5. View History command - (keyword: history)

=== to be added in later PRs ===
5. Replace dummy data with real data mechanism
6. Documentation.
7. Refine GUI rendering of Score

--- Possibly related feature ---
Timer
Scoring
Show quiz history

Cheng20010201 committed:
Modify .gitignore file to ignore /data/*

Cheng20010201 committed:
Change release number in MainApp.java

Cheng20010201 committed:
Add Score class with a defensive manner

Score represents a finished quiz attempt.

Cheng20010201 committed:
Add tests for Score object, especially for the creation method and compareTo() method.

Cheng20010201 committed:
Fix checkstyle issues

Cheng20010201 committed:
Make score test more rigorous.

Cheng20010201 committed:
Add usage of logger into Score

Cheng20010201 committed:
Refactor AB3 terms in Test files.

Cheng20010201 committed:
Refactor AB3 terms in Test files.

Cheng20010201 committed:
Add unit tests and integration test for LocalDatabasePopulator.java

Cheng20010201 committed:
Add unit tests and integration testing to LocalDatabasePopulator.java

xyzhang00 commented on src/test/java/seedu/weeblingo/storage/JsonSerializableFlashcardBookTest.java:
```diff
@@ -6,32 +6,32 @@
-    private static final Path TYPICAL_FLASHCARDS_FILE = TEST_DATA_FOLDER.resolve("typicalFlashcardsBook.json");
-    private static final Path INVALID_FLASHCARD_FILE = TEST_DATA_FOLDER.resolve("invalidFlashcardBook.json");
-    private static final Path DUPLICATE_FLASHCARD_FILE = TEST_DATA_FOLDER.resolve("duplicateFlashcardBook.json");
+            "src", "test", "data", "JsonSerializableAddressBookTest");
```
Is the change back to AddressBookTest intentional? :0

Cheng20010201 replied:
I forgot since the entire test class is ... dummy :>

Will change it back.

Cheng20010201 committed:
Add exception message to JsonDatabaseReader

Cheng20010201 committed:
Refactor address-book into flashcard-book [terminology]

Cheng20010201 committed:
Add isSameAttempt method in Score to check whether the two scores are created at the same time

If two Scores are created with an interval of >= 1 second, they are considered to be created at the same time.

Tests for this method are added as well.

Cheng20010201 committed:
Add DuplicateAttemptScoreException

Essentially a RuntimException. Will be thrown if there are two Scores added to the
list that will be displayed in the GUI.

Cheng20010201 committed:
Add UniqueScoreHistoryList.java

Will be the foundation of displaying score history to user through GUI.

Cheng20010201 committed:
Minor naming convention changes

Cheng20010201 committed:
Fix regressions of testing

Cheng20010201 committed:
Add ScoreList to FlashcardBook, which is the main storage data "transformer"

FlashCardBook::equals and FlashCardBook::hash are not modified since when we compare two FlashcardBooks,
only comparing the content of flashcards suffices.

Cheng20010201 committed:
Update RuntimeException thrown to be self-defined classes
with more description name and documentation

These Exceptions are defined with a defensive programming manner.

Cheng20010201 committed:
Add primitive UI support for ScoreCard

To be beautified by my brilliant teammates.

Cheng20010201 committed:
Refactor AB3 specific term: the list of persons

Cheng20010201 committed:
Add ScoreHistoryListPanel and its fxml support

To be beautified by my wonderful teammates.

Cheng20010201 committed:
Add ScoreHistoryListPanel and its fxml support

To be beautified by my wonderful teammates.

Cheng20010201 committed:
Make the History showing functionality prelimerily work

To be refined after the pending PRs get merged.

Cheng20010201 committed:
Merge branch 'master' into branch-logistic

Cheng20010201 committed:
build correction

Cheng20010201 committed:
Merge pull request #4 from Cheng20010201/master

build correction

Cheng20010201 committed:
Update LocalDatabasePopulator.java

Cheng20010201 committed:
Add JsonAdaptedScore (essentially similar to JsonAdaptedFlashcard)

Cheng20010201 committed:
Update JsonSerializableFlashcardBook.java

Cheng20010201 committed:
Fix minor UI issue

Cheng20010201 committed:
Get back MainWindow to the original state

Cheng20010201 committed:
Add History command and relavant UI rendering.

Design to be refined after the pending PR is merged.

Cheng20010201 committed:
Merge github.com:AY2021S2-CS2103T-T13-1/tp into branch-logistic

# Conflicts:
#	src/main/java/seedu/weeblingo/logic/Logic.java
#	src/main/java/seedu/weeblingo/logic/LogicManager.java
#	src/main/java/seedu/weeblingo/model/Quiz.java
#	src/main/java/seedu/weeblingo/ui/MainWindow.java

Cheng20010201 committed:
The "history" command 'works' with dummy data now

Cheng20010201 committed:
Merge branch 'branch-logistic' of github.com:Cheng20010201/tp into branch-logistic

# Conflicts:
#	src/main/java/seedu/weeblingo/storage/LocalDatabasePopulator.java

Cheng20010201 committed:
Make ListPanel subtitution logic more OOP (not using inheritance/interface)

Cheng20010201 committed:
Merge github.com:AY2021S2-CS2103T-T13-1/tp into branch-logistic

# Conflicts:
#	src/main/java/seedu/weeblingo/storage/LocalDatabasePopulator.java
#	src/main/java/seedu/weeblingo/ui/FlashcardListPanel.java
#	src/main/java/seedu/weeblingo/ui/MainWindow.java

Cheng20010201 committed:
Integrate Score History with Mode; MainWindow.java now looks better

xyzhang00 commented on src/main/java/seedu/weeblingo/storage/LocalDatabasePopulator.java:
```diff
@@ -38,18 +41,48 @@
+
+        try {
+            sampleFb.addScore(Score.of(100, 100));
+            TimeUnit.SECONDS.sleep(1);
```
Is there a reason for calling sleep?

Cheng20010201 replied:
This is the dummy history data generation part. Since (now the `Score` handles the `LocalDatetime` which is the `Score's` creation time internally) && (The precision of `LocalDatetime` is only limited to seconds in our application), a consecutive creation of `Score` object will result in duplicate `Score` objects stored (which, by the same principle of how we store `Flashcard`s, will throw an exception).

When the data is no longer dummy, we can safely assume the user would at least take 1 second between finishing the current quiz and finishing the next quiz.

If the above is not a legitimate assumption, I can change my implementations (but some efforts are required).

xyzhang00 replied:
I think it is a valid assumption! Probably will be better to document it down in the javadoc later.

xyzhang00 commented on src/main/java/seedu/weeblingo/ui/MainWindow.java:
```diff
@@ -215,6 +217,21 @@ private CommandResult executeCommand(String commandText) throws CommandException
     }
 
+    private void changePlaceHolderContent(boolean isShowingHistory) {
+        int index = -1;
```
What is this index referring to? Would it be better to set it as a static variable at the start instead of a magic number?

Cheng20010201 replied:
It is essentially because
```java
int index = -1;
        if (isShowingHistory) {
            index = flashcardListPanelPlaceholder.getChildren().indexOf(flashcardListPanel.getRoot());
            if (index != -1) {
                flashcardListPanelPlaceholder.getChildren().set(index, scoreHistoryListPanel.getRoot());
            }
        }
```
the
```java
indexOf(...)
```
method would return `-1` if the object to look for is not in the `List`.

Thus a static variable seems not very ok but I kind of cannot think of any elegant solution to it...(?:grimacing:)

xyzhang00 replied:
I think there is a .contains() method?

xyzhang00 commented on src/main/java/seedu/weeblingo/ui/ScoreHistoryListPanel.java:
```diff
@@ -0,0 +1,44 @@
+import seedu.weeblingo.commons.core.LogsCenter;
+import seedu.weeblingo.model.score.Score;
+
+public class ScoreHistoryListPanel extends UiPart<Region> implements ListPanel {
```
Is there a reason for these 2 list panels to implement list panel? extending UiPart<Region> seems to be sufficient OOP for me :0

Cheng20010201 replied:
Originally that Interface is to help with a generic method to change the content of `flashcardListPanelPlaceHolder`. Now the current version it is no longer needed. Will fix it thanks!

xyzhang00 reviewed (approved):
Lgtm except for some small queries! But I guess we don't really have to fix them now, probably better to move on with implementation first.

Cheng20010201 committed:
Resolve https://github.com/AY2021S2-CS2103T-T13-1/tp/pull/68#discussion_r603295895

Cheng20010201 committed:
Resolve https://github.com/AY2021S2-CS2103T-T13-1/tp/pull/68#discussion_r600580368
